modules/mod_ginger_rdf/support/sparql_query.erl

```diff
@@ -2,48 +2,100 @@
+    >>.
+
+%% @doc Construct a SPARQL query with a set of predicates.
+-spec select([binary()]) -> sparql_query().
+select(Predicates) ->
+    lists:foldl(
+        fun add_argument/2,
+        #sparql_query{select = [<<"?s">>]},
+        Predicates
+    ).
+
+%% @doc Construct a SPARQL query for a single subject with a set of predicates.
+-spec select(binary(), [binary()]) -> sparql_query().
+select(Uri, Predicates) ->
+    Query = select(Predicates),
+    and_where(<<"VALUES ?s {<", Uri/binary, ">}">>, Query).
 
 %% @doc Resolve query arguments by replacing numerical placeholders with their
 %%      LD predicate counterparts stored in Arguments.
--spec resolve_arguments(Bindings :: map(), arguments()) -> Bindings :: map().
-resolve_arguments(Bindings, Arguments) ->
+-spec resolve_arguments(Bindings :: map(), sparql_query()) -> Bindings :: map().
+resolve_arguments(Bindings, #sparql_query{arguments = Arguments}) ->
     maps:fold(
         fun(Key, Value, Acc) ->
-            ResolvedKey = proplists:get_value(Key, Arguments),
-            Acc#{ResolvedKey => Value}
+            case maps:get(Key, Arguments, undefined) of
+                undefined ->
+                    Acc;
+                ResolvedKey ->
+                    Acc#{ResolvedKey => Value}
+            end
         end,
         #{},
         Bindings
     ).
 
-%% @doc Build a where query section.
--spec where([binary()]) -> {[binary()], [binary()]}.
-where(Properties) ->
-    lists:mapfoldl(
-        fun add_argument/2,
-        [],
-        Properties
-    ).
+%% @doc Add a where clause to a query.
+-spec and_where(binary(), sparql_query()) -> sparql_query().
+and_where(Clause, #sparql_query{where = Clauses} = Query) ->
+    Query#sparql_query{where = [Clause | Clauses]}.
+
+%% @doc Add offset to query.
+-spec offset(non_neg_integer(), sparql_query()) -> sparql_query().
+offset(Offset, Query) when Offset >= 0 ->
+    Query#sparql_query{offset = Offset}.
+
+%% @doc Add limit to query.
+-spec limit(pos_integer(), sparql_query()) -> sparql_query().
+limit(Limit, Query) when Limit > 0 ->
+    Query#sparql_query{limit = Limit}.
 
-%% @doc Add a single property argument.
--spec add_argument(binary(), arguments()) -> {Query :: binary(), arguments()}.
-add_argument(Property, Properties) ->
-    Next = z_convert:to_binary(length(Properties) + 1),
-    {<<"OPTIONAL {?s <", Property/binary, "> ?", Next/binary, "} ">>, Properties ++ [{Next, Property}]}.
+%% @doc Add a query argument.
+-spec add_argument(binary(), sparql_query()) -> sparql_query().
+add_argument(Property, #sparql_query{} = Query) ->
+    #sparql_query{select = Select, arguments = Arguments, where = Where} = Query,
+    Next = z_convert:to_binary(maps:size(Arguments) + 1),
+    Argument = <<"?", Next/binary>>,
+    Query#sparql_query{
+        select = Select ++ [Argument],
+        arguments = Arguments#{Next => Property},
+        where = [<<"OPTIONAL {?s <", Property/binary, "> ?", Next/binary, "}">> | Where]
```

Comment:
(Aside) Am I correct to conclude that predicates are expected to always be in url form, hence the angular braces? E.g. when using schema prefixes the braces should not be there, as in `schema:name` or `a` (built-in predicate, maybe specific to Virtuoso).

Reply:
Tests are unchanged, we'd probably like to have some extra for the query builder.

Reply:
> Am I correct to conclude that predicates are expected to always be in url form

For now, yes. An improvement would to pattern match on `http(s)://` and only then add the guillemets `< >`.